datameta/api/ui/convert.py

```diff
@@ -36,12 +37,35 @@ def formatted_mrec_value(value, datetimefmt):
             pass
     return value
 
+def determine_samplesheet_reader(file_like_obj):
+    # https://readxl.tidyverse.org/reference/excel_format.html
+    XLSX, XLS = b'PK\x03\x04', b'\xd0\xcf\x11\xe0\xa1\xb1\x1a\xe1'
+    nibble = file_like_obj.read(4)
+    def make_excel_reader(sheet):
+        sheet.seek(0)
+        return pd.read_excel(sheet, dtype="object")
+    if nibble == XLSX:
+        reader = make_excel_reader
+    elif nibble + file_like_obj.read(4) == XLS:
+        reader = make_excel_reader
+    else:
+        file_like_obj.seek(0)
+        dialect = csv.Sniffer().sniff(file_like_obj.read(1024).decode())
+        def make_table_reader(sheet, sep=dialect.delimiter):
+            sheet.seek(0)
+            return pd.read_table(sheet, dtype="object", sep=sep)
+        reader = make_table_reader
+
+    return reader
```

Comment:
I've changed this slightly to improve code readability. Also I think it's best to always change the file access position to `0` before returning it and not to rely on the fact that the caller will immediately use the returned func.

```suggestion
def get_samplesheet_reader(file_like_obj):
    """Given a file with tabular data which is either in delimited plain text
    format or in XLS(X) format, returns a function capable of reading the file
    and returning a pandas.DataFrame"""
    # https://readxl.tidyverse.org/reference/excel_format.html
    xlsx_sig   = b'PK\x03\x04', 
    xls_sig    = b'\xd0\xcf\x11\xe0\xa1\xb1\x1a\xe1'
    nibble     = file_like_obj.read(8)

    file_like_obj.seek(0)

    if nibble.startswith(xlsx_sig) or nibble.startswith(xls_sig):
        return lambda sheet : pd.read_excel(sheet, dtype="object")
    else:
        dialect = csv.Sniffer().sniff(file_like_obj.read(1024).decode())
        file_like_obj.seek(0)
        return lambda sheet : pd.read_table(sheet, dtype="object", sep=dialect.delimiter)
```

Reply:
@lkuchenb you mean lambdas are more readable than closures? :P

Reply:
Nothing I'd be religious about, but after moving the seek into the main function scope the functions became one-liners ;-)

Reply:
Fair enough. I kept the closures as I think they're easier to read than the lambdas (prefer lambdas in map/filter context).

Reply:
Fine with me 😄